### ISSUE TITLE
[feature-request] Require first-party imports to target highest exporting sibling-or-lower exporting module

### ISSUE THREAD

**Issue by smackesey** [charliermarsh/ruff#1107]
A symbol can be defined in one module and re-exported from several others. When importing a symbol from an external typed library, the convention is to import from as close to the root as possible. This is how popular auto-import resolution tools work (see my very similar post in [this pylance discussion](https://github.com/microsoft/pylance-release/discussions/3050#discussioncomment-3197660): 

>If the same name is exported from multiple modules, the auto-import logic in pyright and pylance prefer the shortest module path. Only the shortest path is listed in the completion list, and the longer paths are de-duped.

In large projects, sometimes you have big submodules that you want to provide an "internal public" interface for-- e.g. if you have a large submodule `project.foo`, you might want project-internal code to import only from `project.foo` itself, rather than `project.foo.*` submodules.

I'd like to suggest that `ruff` offer a rule that enforces this by requiring internal auto-imports instead resolve to the shortest non-ancestor module that exposes a symbol. So:

```
### project/foo/__init__.py
# redundant alias is necessary to expose `BAR` as public for `project.foo`
from .bar import BAR as BAR

### project/foo/bar.py
BAR = "BAR"

### project/baz.py
from project.foo.bar import BAR   # ERROR: should import from project.foo instead
```

**Comment by charliermarsh** [comment]
Yeah this makes sense. It would also force us to support cross-module checks (since we'd need to build a mapping from module to exported symbols), which would be a good thing.

**Comment by chbndrhnns** [comment]
I sometimes see modules prefixed with an underscore being treated as private. Here is an example how this could work:

```py
# ./client.py
from .myp._private import Member
from .myp._private import OtherMember

assert Member

# ./myp/__init__.py
__all__ = ["Member"]

from ._private import Member

# ./myp/_private.py
class Member: ...

class OtherMember: ...
```

Ruff could turn this into:

```py
# ./client.py
from .myp import Member
from .myp._private import OtherMember

assert Member

# ./myp/__init__.py
__all__ = ["Member"]

from ._private import Member

# ./myp/_private.py
class Member: ...

class OtherMember: ...
```

**Comment by smackesey** [comment]
Just wondering whether this is now easily doable with all the changes in the last year to ruff. Would be a nice QOL improvement for us.
